Publish spec for sample apps (#155)
aws-observability/aws-otel-community: sample-apps/SampleAppSpec.md

> ## Standardized Sample App Spec - README

Author: Ok will do

### Top level Requirements

#### Where the Sample Apps should live

Each sample app that follows the requirements below should be stored in the `sample-apps` folder in https://github.com/aws-observability/aws-otel-community.

> The matrix that exists in the README file in that directory should be updated with the sample apps in conformance to this document. The matrix must additionally display the status of Metrics Logs and Traces. (Stable or Unstable)

Reviewer: ```suggestion
The README file in the `sample-apps` directory contains a matrix which lists the sample apps in conformance to this document. New sample apps must be added to this matrix. The matrix must display the status of Metrics Logs and Traces (Stable or Unstable).
```

#### How the Sample Apps should work

> Each sample app should emit the same Metrics, Logs and Traces. If a sample app cannot emit the mentioned, then that will noted in that sample app’s documentation.

Reviewer: ```suggestion
Each sample app must emit the same Metrics, Logs and Traces. If a sample app cannot emit the mentioned, then that will noted in that sample app’s documentation.
```
Should or must?

Author: I think must would better enforce the standard

> Each sample app should emit the same Metrics, Logs and Traces. If a sample app cannot emit the mentioned, then that will noted in that sample app’s documentation.

Reviewer: ```suggestion
Each sample app should emit the same Metrics, Logs and Traces according to this specification. The sample app README should list any missing signal type in it's implementation. 
```

Each Sample app should should be able to be built as an image. The sample app must be able to accept configuration through a configuration file (YAML) and environment variables.

#### Assumption on how the Sample Apps should emit Traces and Metrics

**Traces** are not something that could or should be generated arbitrarily but instead should reflect a real life cycle or flow.

> In existing sample applications, Trace instrumentation is coupled with a **Traffic Generator** (https://github.com/aws-observability/aws-otel-test-framework/tree/terraform/traffic-generator) that makes http requests to the Sample app and emits traces based on that those requests. In another example, **ho11y** (https://gitlab.aws.dev/hausenbl/); sample apps can create their own traffic by invoking other sample apps (invocation pattern). 

Reviewer: ```suggestion
In existing sample applications, Trace instrumentation is coupled with a **Traffic Generator** (https://github.com/aws-observability/aws-otel-test-framework/tree/terraform/traffic-generator) that makes http requests to the sample app and emits traces based on those requests. In another example, **ho11y** (https://gitlab.aws.dev/hausenbl/); sample apps can create their own traffic by invoking other sample apps (invocation pattern). 
```

Sample apps in conformance to this document will also include the invocation pattern as an optional configuration. This will allow for more extensive testing and demos if configured.
A possible example can be seen below. This invocation capability can allow testing to be as complex or simple as needed.

![Alt text](./invocation.png?raw=true "Invocation")

The two patterns of metrics that will be generated are **Request-based** and **Random**. Request based metrics will also assume an invocation has been made to the Sample App. Random can be arbitrarily generated for metrics that do not correlate to incoming requests. 

> ### Documentation Requirements

Author: Reworked to include the description, non-conformance, and workarounds within the README.

Each sample app MUST be coupled with a document that will have the following sections:

* **Description** - Information about the language being used and a link to this document for context
* **Non-conformance** - Documentation on missing features or missing extensions that don’t allow a given sample app to conform with this document
* **Workarounds** - Examples of workarounds include the use of an unstable API or SDK to produce signals. This code `workaround` code may require changes in the future.
* **README** - The readme that will go into each sample app’s code base detailing how to setup the sample app (this should be same across all sample apps); although redundant, each repository should still have it to avoid looking across sample apps for a readme. This readme should also include the Non-conformance section specific to the sample app it belongs to.

### Functional Requirements

Each sample app should be configured with manual instrumentation from OTel. Auto instrumentation can also be implemented.

> Metrics and traces should be OTLP and Logs should be in JSON format using any popular logging library for each respective language with the appropriate license. (e.g. Log4j for java, Logrus for Go)

Author: Yes it seems to be referring to the use of the OTLP grpc exporter: https://github.com/open-telemetry/opentelemetry-collector/blob/main/exporter/otlpexporter/README.md.  It seems like Logs are in beta for this exporter so should we use it for logs as well so that all signals use the OTLP exporter?

> Metrics and traces should be OTLP and Logs should be in JSON format using any popular logging library for each respective language with the appropriate license. (e.g. Log4j for java, Logrus for Go)

Reviewer: I think that would make sense

> Metrics and traces should be OTLP and Logs should be in JSON format using any popular logging library for each respective language with the appropriate license. (e.g. Log4j for java, Logrus for Go)

Author: Ok I have updated the Spec to reflect that all signals will use the OTLP exporter

This configuration file should be set by an environment variable (SAMPLE_APP_CONF) or else it will default to the below snippet. If two sample apps are deployed by default, there will be an error for using the same port twice.

```
---
Host: "0.0.0.0"                       # Host - String Address
Port: "8080"                          # Port - String Port
TimeInterval: 1                       # Interval - Time in seconds to generate new metrics
RandomTimeAliveIncrementer: 1         # Metric - Amount to incremement metric by every TimeInterval
RandomTotalHeapSizeUpperBound: 100    # Metric - UpperBound for TotalHeapSize for random metric value every TimeInterval
RandomThreadsActiveUpperBound: 10     # Metric - UpperBound for ThreadsActive for random metric value every TimeInterval
RandomCpuUsageUpperBound: 100         # Metric - UppperBound for CpuUsage for random metric value every TimeInterval
SampleAppPorts: []                    # Sampleapp ports to make calls to   
ResourceDetector: ''                  # String to specify resource detector
```

Every sample app will assume a configuration file with these variable names.

> The testing framework MUST not test against these Metrics due to upstream unstable SDK & API in certain languages. These are mainly for being able to demonstrate some type of metrics being generated in languages that support it. 

Author: Ok so the test framework should assert the existence of these metrics for all sample apps but only do metric validations for the those sdks that are stable in metrics?

> The testing framework MUST not test against these Metrics due to upstream unstable SDK & API in certain languages. These are mainly for being able to demonstrate some type of metrics being generated in languages that support it. 

Reviewer: The test framework doesn't generally validate against the value of metrics, just their existence. This reads like the test framework should not perform any tests against metrics from the sample apps.

> The testing framework MUST not test against these Metrics due to upstream unstable SDK & API in certain languages. These are mainly for being able to demonstrate some type of metrics being generated in languages that support it. 

Reviewer: I think this wording still needs to be updated. The testing framework can run validations against these metrics.

## Functional Requirements

Each sample app should be configured with manual instrumentation from OTel. Auto instrumentation can also be implemented.

> Metrics, traces, and logs will all use an OTLP exporter to be exported from the sameple app in OTLP/JSON format.

Reviewer: ```suggestion
Metrics, traces, and logs will all use an OTLP exporter to ship data from the sample app.
```
